workspaces/vscode-python/Dockerfile

```diff
@@ -8,38 +8,34 @@ ARG CACHEBUST=2025-12-15-14-21-51
 # installed, as a separate step below.
 SHELL ["/bin/bash", "-lc"]
 
-# Install Python and required packages via miniconda.
+# Ensure that running Python in the container will use the correct Python version.
+ENV PATH="/home/coder/.venv/bin:$PATH"
+
+# - Ensure that all `uv` commands compile Python source files to bytecode.
+# - Ensure that all `uv` commands do not use any caching.
+ENV UV_COMPILE_BYTECODE=1 UV_NO_CACHE=1
+
+USER root
+
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends gcc libc6-dev graphviz graphviz-dev \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
 # This is a prerequisite for the VS Code extensions in the following steps.
 USER coder
+WORKDIR "/home/coder"
+
+# Install Python and required packages via uv.
+# /home/coder/.venv/bin/python3 -> /home/coder/.local/bin/python3 -> /home/coder/.local/share/uv/python/*/bin/python3.10
 COPY requirements.txt /
-# SC1091: `source` an unknown file is required to activate Conda.
-# hadolint ignore=SC1091
-RUN arch="$(uname -m)" \
-    && curl -sfLO "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-${arch}.sh" \
-    && chmod +x "Miniforge3-Linux-${arch}.sh" \
-    && ./"Miniforge3-Linux-${arch}.sh" -b -p /home/coder/conda \
-    # Install conda and mamba hooks for future interactive bash sessions:
-    && /home/coder/conda/bin/conda init bash \
-    # Activate hooks in the current noninteractive session:
-    && . "/home/coder/conda/etc/profile.d/conda.sh" \
-    && . "/home/coder/conda/etc/profile.d/mamba.sh" \
-    && mamba activate \
-    # Installing `pygraphviz` with pip would require `build-essentials`, `graphviz`,
-    # and `graphviz-dev` to be installed at the OS level, which would increase the
-    # image size. Instead, we install it from Conda, which prebuilds it and also
-    # automatically installs a Conda-specific `graphviz` dependency.
-    && mamba install --yes "$(grep pygraphviz /requirements.txt | head -n 1)" \
-    && pip install --no-cache-dir -r /requirements.txt \
-    && rm "Miniforge3-Linux-${arch}.sh" \
-    && mamba clean --all --yes --quiet \
-    && pip cache purge
-
-# Also add conda init to /etc/profile.d/ so that noninteractive login shells
-# can use "pip install" (e.g. during additional Docker RUN commands). This is
-# necessary because the bash hook installed with "init bash" above will only
-# activate for interactive shells.
-USER 0
-RUN /home/coder/conda/bin/conda init --system
+RUN curl -LO https://astral.sh/uv/install.sh \
+    && sh install.sh && rm install.sh \
+    && source "$HOME/.local/bin/env" \
+    && uv python install --default 3.12 \
+    && uv venv \
+    && echo "source $HOME/.venv/bin/activate" >> ~/.bashrc && source "$HOME/.venv/bin/activate" \
+    && uv pip install --no-cache-dir -r /requirements.txt
```

Comment:
The C grader image uses the following order:

- Create `venv`
- Copy in `requirements.txt`
- Install dependencies from `requirements.txt`.

I'd be inclined to do the same here, it'll make better use of Docker layer caching.

Reply:
Done in [dfb7414](https://github.com/PrairieLearn/PrairieLearn/pull/13520/commits/dfb7414ef38e1ad06385037ba5e6eaabbc2d8e6f)

```
❯ docker run -it --entrypoint /bin/sh prairielearn/workspace-vscode-python
$ whoami
coder
$ which python3
/home/coder/.venv/bin/python3
$ python3
Python 3.12.12 (main, Dec  9 2025, 19:02:07) [Clang 21.1.4 ] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> ^D
$ echo $PATH
/home/coder/.venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```